lib/Sema/CSSimplify.cpp

```diff
@@ -6951,6 +6963,32 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
     return matchTypes(type1, type2, matchKind, subflags, locator);
   }
 
+  case FixKind::AllowTupleTypeMismatch: {
+    auto lhs = dyn_cast<TupleType>(type1.getPointer());
+    auto rhs = dyn_cast<TupleType>(type2.getPointer());
+    auto lhsLarger = lhs->getElements().size() >= rhs->getElements().size();
+    auto larger = lhsLarger ? lhs : rhs;
+    auto smaller = lhsLarger ? rhs : lhs;
+    if (lhs && rhs && getContextualTypePurpose() == CTP_Initialization) {
+      // Match up the tuple type elements, and match any
+      for (unsigned i = 0; i < larger->getElements().size(); ++i) {
+        auto largerTy = larger->getElement(i).getType();
+        if (i < smaller->getElements().size()) {
+          auto smallerTy = smaller->getElement(i).getType();
+          if (smallerTy->isTypeVariableOrMember() ||
+              largerTy->isTypeVariableOrMember())
+            addConstraint(ConstraintKind::Bind, largerTy, smallerTy,
```

Comment:
In case of arity mismatch, instead of doing matching manually it would be better to create a new tuple with correct number of arguments and use `matchTupleTypes` (type variables, defaultable to `Any` could be used to feel missing elements). Also in certain conditions, e.g. generic requirements, types have to match exactly so elements has to be equated via `Equal` constraint but in most cases conversion is allowed.